### PR TITLE
Deprecate `extraction_operator_linter()`

### LINKED ISSUE #1484: False positive in `extraction_operator_linter()` with plotmath expressions
One can use plotmath to display mathematical expressions in a graphic:

``` r
library(ggplot2)
expr <- list(quote(italic("x")["i"]))
ggplot() + labs(title = parse(text = expr))
```

<sup>Created on 2022-07-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

But the code for subscript produces an incorrect lint:

``` r
library(lintr)

lint(
  text = "list(quote(italic('x')['i']))",
  linters = extraction_operator_linter()
)
#> <text>:1:23: warning: [extraction_operator_linter] Use `[[` instead of `[` to extract an element.
#> list(quote(italic('x')['i']))
#>                       ^
```

<sup>Created on 2022-07-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

The suggestion is misleading as it won't produce the same output:

``` r
library(ggplot2)
expr <- list(quote(italic("x")[["i"]]))
ggplot() + labs(title = parse(text = expr))
```

<sup>Created on 2022-07-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### LINKED ISSUE #1485: False positive in `extraction_operator_linter()` with `R6` class method calls
This shouldn't produce a lint:

``` r
library(lintr)

lint(
  text = "Person <- R6Class('Person', list(
          name = NULL,
          initialize = function(name) self$name <- name
          ))
  
  Bob <- Person$new('Bob')",
  linters = extraction_operator_linter()
)
#> <text>:6:16: warning: [extraction_operator_linter] Use `[[` instead of `$` to extract an element.
#>   Bob <- Person$new('Bob')
#>                ^
```

<sup>Created on 2022-07-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

And, although the suggested replacement would work here, this would be an extremely unusual way of calling `R6` class methods:

``` r
library(R6)

Person <- R6Class('Person', list(
  name = NULL,
  initialize = function(name) self$name <- name
))

Person[["new"]]('Bob')
#> <Person>
#>   Public:
#>     clone: function (deep = FALSE) 
#>     initialize: function (name) 
#>     name: Bob
```

<sup>Created on 2022-07-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

### PR DESCRIPTION
**Pull request by IndrajeetPatil**
Closes #1485
Closes #1484

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -12,6 +12,7 @@
    + Helper `with_defaults()`.
 * `all_linters()` has signature `all_linters(..., packages)` rather than `all_linters(packages, ...)` (#2332, @MichaelChirico). This forces `packages=` to be supplied by name and will break users who rely on supplying `packages=` positionally, of which we found none searching GitHub.
 * Adjusted various lint messages for consistency in readability (#1330, @MichaelChirico). In general, we favor lint messages to be phrased like "Action, reason" to but the "what" piece of the message front-and-center. This may be a breaking change for code that tests the specific phrasing of lints.
+* `extraction_operator_linter()` is deprecated. Although switching from `$` to `[[` has microoptimization benefits, it can lead to non-idiomatic code in many contexts (e.g. R6 classes, Shiny applications, etc.) (#2409, @IndrajeetPatil).
```
Comment (MichaelChirico) [inline NEWS.md:15]:
```suggestion
* `extraction_operator_linter()` is deprecated. Although switching from `$` to `[[` has some robustness benefits for package code, it can lead to non-idiomatic code in many contexts (e.g. R6 classes, Shiny applications, etc.) (#2409, @IndrajeetPatil).
```

Comment (MichaelChirico) [reply NEWS.md:15]: maybe also mention the warnPartialMatchDollar (sp?) option here too as a different way to prevent issues related to partial matching with $.

**Review by MichaelChirico — APPROVED** [review]
thanks!

**Comment by IndrajeetPatil** [comment]
Ah, there are new warnings detected by the vigilant workflow that need to be dealt with before we can merge this.

**Comment by IndrajeetPatil** [comment]
I don't understand what's going on here. 

In the `dummy_packages/package` we are using in `test-cache.R` (e.g.), we don't specify `extraction_operator_linter` in the config and just set `linters: linters_with_defaults()`, so why do we see the following warning?

> Trying to remove 'extraction_operator_linter', which is not in `defaults`.

Full traceback: 

```r
── Warning (test-cache.R:446:3): cache = TRUE workflow works ───────────────────
Warning encountered while loading config:
  Warning from config setting 'linters' in 'NULL':
    Trying to remove 'extraction_operator_linter', which is not in `defaults`.
Backtrace:
     ▆
  1. ├─lintr::lint_package(pkg, cache = TRUE) at test-cache.R:446:3
  2. │ └─lintr:::read_settings(pkg_path) at lintr/R/lint.R:249:5
  3. │   └─lintr:::read_config_file(config_file) at lintr/R/settings.R:74:3
  4. │     ├─base::withCallingHandlers(...) at lintr/R/settings.R:140:3
  5. │     ├─base::tryCatch(load_config(config_file), error = malformed)
  6. │     │ └─base (local) tryCatchList(expr, classes, parentenv, handlers)
  7. │     │   └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
  8. │     │     └─base (local) doTryCatch(return(expr), name, parentenv, handler)
  9. │     └─lintr (local) load_config(config_file)
 10. │       ├─base::withCallingHandlers(...) at lintr/R/settings.R:113:9
 11. │       ├─base::tryCatch(...)
 12. │       │ └─base (local) tryCatchList(expr, classes, parentenv, handlers)
 13. │       │   └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
 14. │       │     └─base (local) doTryCatch(return(expr), name, parentenv, handler)
 15. │       ├─base::eval(parsed_setting)
 16. │       │ └─base::eval(parsed_setting)
 17. │       └─lintr::all_linters(...)
 18. │         └─lintr::linters_with_tags(tags = NULL, packages = packages, ...) at lintr/R/with.R:141:3
 19. │           └─lintr::modify_defaults(..., defaults = tagged_linters) at lintr/R/with.R:124:3
 20. │             └─base::warning(...) at lintr/R/with.R:48:5
 21. ├─base::.signalSimpleWarning(...)
 22. │ └─base::withRestarts(...)
 23. │   └─base (local) withOneRestart(expr, restarts[[1L]])
 24. │     └─base (local) doWithOneRestart(return(expr), restart)
 25. ├─lintr (local) `<fn>`(`<smplWrnn>`)
 26. │ └─base::warning(...) at lintr/R/settings.R:125:13
 27. ├─base::.signalSimpleWarning(...)
 28. │ └─base::withRestarts(...)
 29. │   └─base (local) withOneRestart(expr, restarts[[1L]])
 30. │     └─base (local) doWithOneRestart(return(expr), restart)
 31. └─lintr (local) `<fn>`(`<smplWrnn>`)
```

**Comment by AshesITR** [comment]
Traceback no 17. shows that it runs `all_linters()`. Seems to pick up our config, where the warning is correct because `all_linters()` excludes deprecated and defunct linters by default.

TK;DR: Remove `extraction_operator_linter = NULL` from our `.lintr`.

**Comment by IndrajeetPatil** [comment]
> TL;DR: Remove `extraction_operator_linter = NULL` from our `.lintr`.

This is what I had thought, but then I realized that this is going to cause the lint workflow to fail IINM. 
I hope we don't have to add `# nolint` directives to everywhere `$` is appearing.

**Review by MichaelChirico — APPROVED** [review]
